src/pybind/mgr/telemetry/module.py

```diff
@@ -0,0 +1,355 @@
+
+        if option == 'interval':
+            try:
+                value = int(value)
+            except (ValueError, TypeError):
+                raise RuntimeError('invalid interval. Please provide a valid '
+                                   'integer')
+
+            if value < 24:
+                raise RuntimeError('interval should be set to at least 24 hours')
+
+        if option in ['leaderboard', 'enabled']:
+            value = self.str_to_bool(value)
+
+        if option == 'contact':
+            if value and not self.is_valid_email(value):
+                raise RuntimeError('%s is not a valid e-mail address as a '
+                                   'contact', value)
+
+        if option in ['description', 'organization']:
+            if value and len(value) > 256:
+                raise RuntimeError('%s should be limited to 256 '
+                                   'characters', option)
+
+        self.config[option] = value
+        return True
+
+    def init_module_config(self):
+        for key, default in self.config_keys.items():
+            self.set_config_option(key, self.get_config(key, default))
+
+        self.last_upload = self.get_store('last_upload', None)
+        if self.last_upload is not None:
+            self.last_upload = int(self.last_upload)
+
+        self.report_id = self.get_store('report_id', None)
+        if self.report_id is None:
+            self.report_id = str(uuid.uuid4())
+            self.set_store('report_id', self.report_id)
+
+    def gather_osd_metadata(self, osd_map):
+        keys = ["osd_objectstore", "rotational"]
+        keys += self.metadata_keys
+
+        metadata = dict()
+        for key in keys:
+            metadata[key] = defaultdict(int)
+
+        for osd in osd_map['osds']:
+            for k, v in self.get_metadata('osd', str(osd['osd'])).items():
+                if k not in keys:
+                    continue
+
+                metadata[k][v] += 1
+
+        return metadata
+
+    def gather_mon_metadata(self, mon_map):
+        keys = list()
+        keys += self.metadata_keys
+
+        metadata = dict()
+        for key in keys:
+            metadata[key] = defaultdict(int)
+
+        for mon in mon_map['mons']:
+            for k, v in self.get_metadata('mon', mon['name']).items():
+                if k not in keys:
+                    continue
+
+                metadata[k][v] += 1
+
+        return metadata
+
+    def compile_report(self):
+        report = {'leaderboard': False, 'report_version': 1}
+
+        if self.str_to_bool(self.config['leaderboard']):
+            report['leaderboard'] = True
+
+        for option in ['description', 'contact', 'organization']:
+            report[option] = None
+            if self.config[option]:
+                report[option] = self.config[option]
+
+        mon_map = self.get('mon_map')
+        osd_map = self.get('osd_map')
+        fs_map = self.get('fs_map')
+        df = self.get('df')
```

Comment:
could you dump some pool stats in the reports as well? as Sage suggested in the mailing list? and the service_map?

Reply:
I'd have to look that up. At the moment I'm very limited on time so I'd rather have this module in before the Mimic release so that it's at least up there.

I will look into this later.